dmoj/cptbox/utils.py

```diff
@@ -1,11 +1,85 @@
+import errno
 import io
+import mmap
+import os
+from tempfile import NamedTemporaryFile
+from typing import Optional
 
 from dmoj.cptbox._cptbox import memory_fd_create, memory_fd_seal
 
 
-class MemoryIO(io.FileIO):
-    def __init__(self) -> None:
-        super().__init__(memory_fd_create(), 'r+')
+def _make_fd_readonly(fd):
```

Comment:
This can fail silently.

Reply:
What do you mean? If `os.open` or `os.dup2` fails, `OSError` is raised.